include/deal.II/matrix_free/fe_point_evaluation.h

```diff
@@ -1729,17 +1971,160 @@ FEPointEvaluation<n_components_, dim, spacedim, Number>::evaluate(
+              gradient_in_face[0] = gradient[0];
+              gradient_in_face[1] = gradient[1];
+            }
+          else
+            Assert(false, ExcInternalError());
+        }
+      else
+        Assert(false, ExcInternalError());
+
+      internal::integrate_tensor_product_value_and_gradient<
+        dim - 1,
+        VectorizedArrayType,
+        vectorized_value_type,
+        2>(shapes_faces.data() + qb * n_shapes,
+           n_shapes,
+           value_face.data(),
+           gradient_in_face,
+           solution_renumbered_vectorized.data(),
+           unit_point_faces_ptr[qb],
+           polynomials_are_hat_functions,
+           qb != 0);
+    }
+  else
+    {
+      internal::integrate_tensor_product_value_and_gradient<
+        dim,
+        VectorizedArrayType,
+        vectorized_value_type>(shapes.data() + qb * n_shapes,
+                               n_shapes,
+                               &value,
+                               gradient,
+                               solution_renumbered_vectorized.data(),
+                               unit_point_ptr[qb],
+                               polynomials_are_hat_functions,
+                               qb != 0);
+    }
+}
+
+
+
+template <int n_components_, int dim, int spacedim, typename Number>
+inline void
+FEPointEvaluation<n_components_, dim, spacedim, Number>::finish_integrate_fast(
+  const ArrayView<ScalarNumber> &         solution_values,
+  const EvaluationFlags::EvaluationFlags &integration_flags)
+{
+  std::fill(solution_values.begin(), solution_values.end(), ScalarNumber());
+  for (unsigned int comp = 0; comp < n_components; ++comp)
+    {
+      if (use_face_path)
+        {
+          const unsigned int   size_input = 3 * dofs_per_component_face;
+          VectorizedArrayType *input      = scratch_data_vectorized.begin();
+          VectorizedArrayType *output     = input + size_input;
+
+          for (unsigned int i = 0; i < 3 * dofs_per_component_face; ++i)
+            ETT::write_value(input[i], comp, solution_renumbered_vectorized[i]);
+
+          internal::FEFaceNormalEvaluationImpl<dim, -1, VectorizedArrayType>::
+            template interpolate<false, false>(1,
+                                               integration_flags,
+                                               shape_info_vectorized,
+                                               input,
+                                               output,
+                                               current_face_number);
+
+          for (unsigned int i = 0; i < dofs_per_component; ++i)
+            {
+              VectorizedArrayType &result = output[i];
+              for (unsigned int lane = n_lanes_internal / 2; lane > 0;
+                   lane /= 2)
+                for (unsigned int j = 0; j < lane; ++j)
+                  result[j] += result[lane + j];
+              solution_values[renumber[comp * dofs_per_component + i]] =
+                result[0];
+            }
```

Comment:
Let me post this comment separately: Wouldn't it make sense to run the reduction across the vector lanes first, and only afterwards do `FEFaceNormalEvaluationImpl`? The reason why I'm asking is that the additions would need to be done on fewer points when done on the the face for values and normal derivatives (assuming the degree is 2 or larger). That would also avoid using `shape_info_vectorized`, so an alternative to my comment above.

Reply:
Good comment! I tried it out and it runs even a bit faster. See commit 79b2f07cb211b79fe437f3c38d2a2cbedb67b106